### PR TITLE
Enable arbitrary slot ids in DB specs

### PR DESCRIPTION
**Pull request by KtorZ**
# Issue Number

#644

# Overview

- [x] I have reworked a bit how the 'Arbitrary' DB instances were organized to make things a bit more readable
- [x] I have fixed an issue with the DB model (when rolling back beyond the first available checkpoint)
- [x] I have change the arbitrary generation of checkpoints to use arbitrary slot ids (so far, we were only generating checkpoints at slot 0 0!! Not so ideal to test rollbacks and behavior with multiple checkpoints....)

# Comments

The whole diff might be a bit hard to read because one of the commit is moving a lot of code around. I suggest to have a look at it commit by commit, which should make things more obvious!

### CODE REVIEW

File: lib/core/test/shared/Cardano/Wallet/DummyTarget/Primitive/Types.hs [hunk 1]
```diff
@@ -80,16 +79,12 @@ instance PersistTx DummyTarget where
...
 instance KeyToAddress DummyTarget RndKey where
-    keyToAddress k = Address
-        $ CBOR.toStrictByteString
-        $ CBOR.encodeAddress (getRawKey k)
-            [ CBOR.encodeDerivationPathAttr pwd acctIx addrIx ]
-      where
-        (acctIx, addrIx) = derivationPath k
-        pwd = payloadPassphrase k
+    keyToAddress =
+        Address . BS.take 8 . convertToBase Base16 . unXPub . getRawKey
```
Comment (KtorZ) [inline lib/core/test/shared/Cardano/Wallet/DummyTarget/Primitive/Types.hs]: Converting to base16 and using only the first 8 bytes for more readability in test failures! 
At the DB-level, addresses are really just a bunch of bytes, so no need to maintain any sort of CBOR structure there.

File: lib/core/test/unit/Cardano/Wallet/DB/Arbitrary.hs [hunk 2]
```diff
@@ -319,161 +329,189 @@ instance Arbitrary SlotId where
...
+-- We would expect the latest checkpoint to be the one we just inserted, but
+-- currently, it'll also contains the pending indexes from the previously
+-- inserted checkpoints!
+--
+--    shrink =
+--        map Seq.pendingIxsFromList . shrink . Seq.pendingIxsToList
+--    arbitrary =
+--        Seq.pendingIxsFromList . Set.toList <$> arbitrary
+instance Arbitrary (Seq.PendingIxs) where
+    arbitrary = pure Seq.emptyPendingIxs
```
Comment (KtorZ) [inline lib/core/test/unit/Cardano/Wallet/DB/Arbitrary.hs:447]: This is the only change in logic in this file (rest is just functions declarations being moved around). 
We need to review how pending indexes are stored before we can enable arbitrary pending indexes in tests. Will do in another PR.

File: lib/core/test/unit/Cardano/Wallet/DB/Properties.hs [hunk 3]
```diff
@@ -250,7 +266,13 @@ sortedUnions = map (Identity . sort' . runIdentity) . unions
...
 -- | Execute an action once per key @k@ present in the given list
 once :: (Ord k, Monad m) => [(k,v)] -> ((k,v) -> m a) -> m [a]
-once xs = forM (Map.toList (Map.fromList xs))
+once xs action = fmap catMaybes $ flip evalStateT mempty $
+    forM xs $ \(k, v) -> do
+        s <- get
+        modify' (Set.insert k)
+        case Set.member k s of
+            False -> Just <$> lift (action (k, v))
+            True  -> pure Nothing
```
Comment (KtorZ) [inline lib/core/test/unit/Cardano/Wallet/DB/Properties.hs]: I've changed the implementation to not use `Map` because, by doing `Map.toList` and `Map.fromList`, we were loosing some possible ordering of the keys. Now using a set instead to check if the action has already been executed.

File: lib/core/test/unit/Cardano/Wallet/DB/Properties.hs [hunk 4]
```diff
@@ -363,13 +385,16 @@ prop_readAfterPut putOp readOp db (ShowFmt key, ShowFmt a) =
...
+        run $ cleanDB db
+        (InitialCheckpoint cp, meta) <- namedPick "Initial Checkpoint" arbitrary
+        run $ unsafeRunExceptT $ createWallet db key cp meta mempty
+    prop = do
+        run $ unsafeRunExceptT $ putOp db key a
+        res <- run $ readOp db key
+        let fa = pure a
+        monitor $ counterexample $ "\nInserted\n" <> pretty fa
+        monitor $ counterexample $ "\nRead\n" <> pretty res
+        assertWith "Inserted == Read" (res == fa)
```
Comment (KtorZ) [inline lib/core/test/unit/Cardano/Wallet/DB/Properties.hs:397]: reworked using `monitor` and `run` to get better counter examples printed on failures. Did this for debugging and kept it as it produces a nicer failure.

File: lib/core/test/unit/Cardano/Wallet/DB/Arbitrary.hs [hunk 5]
```diff
@@ -160,155 +159,166 @@ import qualified Data.ByteString as BS
...
-instance Arbitrary (PrimaryKey WalletId) where
-    shrink _ = []
+-- | For checkpoints, we make sure to generate them in order.
+instance {-# OVERLAPS #-} (Arbitrary k, Ord k, GenState s)
+    => Arbitrary (KeyValPairs k (ShowFmt (Wallet s DummyTarget))) where
+    shrink = genericShrink
     arbitrary = do
-        bytes <- B8.pack . pure <$> elements ['a'..'k']
-        return $ PrimaryKey $ WalletId $ hash bytes
+        pairs <- choose (1, 10) >>= flip vectorOf arbitrary
```
Comment (paweljakubas) [inline lib/core/test/unit/Cardano/Wallet/DB/Arbitrary.hs]: this means we are about to generate maximum 10 checkpoints?

File: lib/core/test/unit/Cardano/Wallet/DB/Arbitrary.hs [hunk 6]
```diff
@@ -160,155 +159,166 @@ import qualified Data.ByteString as BS
...
+    shrink (MockChain chain) =
+        [ MockChain chain'
+        | chain' <- shrinkList shrinkBlock chain
+        , not (null chain')
+        ]
+      where
+        shrinkBlock (Block h txs) = Block h <$> shrinkList shrink txs
     arbitrary = do
-        SeqState <$> arbitrary <*> arbitrary <*> arbitrary
+        n0 <- choose (1, 10)
```
Comment (paweljakubas) [inline lib/core/test/unit/Cardano/Wallet/DB/Arbitrary.hs:228]: this means we are about to generate maximum 10 checkpoints?

File: lib/core/test/unit/Cardano/Wallet/DB/Arbitrary.hs [hunk 7]
```diff
@@ -160,155 +159,166 @@ import qualified Data.ByteString as BS
...
+                                   Blocks
+-------------------------------------------------------------------------------}
 
-instance Arbitrary TxStatus where
-    arbitrary =
-        elements [Pending, InLedger, Invalidated]
+instance Arbitrary BlockHeader where
+    arbitrary = do
+        sid@(SlotId ep sl) <- arbitrary
+        let h = fromIntegral sl * fromIntegral len + fromIntegral ep * 100
```
Comment (paweljakubas) [inline lib/core/test/unit/Cardano/Wallet/DB/Arbitrary.hs]: why 100? maybe it is good to give a name to 100 ?

File: lib/core/src/Cardano/Wallet/DB/Model.hs [hunk 8]
```diff
@@ -218,13 +224,17 @@ mRollbackTo wid point db = flip (alterModel wid) db $ \wal ->
...
-    findNearestPoint :: [Wallet s t] -> SlotId
-    findNearestPoint = head . sortOn Down . mapMaybe fn
+    -- | Find nearest checkpoint's slot before or equal to 'point'.
+    findNearestPoint :: [Wallet s t] -> Maybe SlotId
+    findNearestPoint = headMay . sortOn Down . mapMaybe fn
       where
         fn :: Wallet s t -> Maybe SlotId
         fn cp = if (tip cp <= point) then Just (tip cp) else Nothing
 
+    headMay :: [a] -> Maybe a
```
Comment (paweljakubas) [inline lib/core/src/Cardano/Wallet/DB/Model.hs]: why not name it `headSafe` or `headM` :-)

Comment (KtorZ) [reply lib/core/src/Cardano/Wallet/DB/Model.hs]: `headSafe` sounds good :) 
The `M` suffix is often used for `Monad` though, so I wouldn't use it here.

File: lib/core/test/unit/Cardano/Wallet/DB/Properties.hs [hunk 9]
```diff
@@ -250,7 +266,13 @@ sortedUnions = map (Identity . sort' . runIdentity) . unions
 
 -- | Execute an action once per key @k@ present in the given list
 once :: (Ord k, Monad m) => [(k,v)] -> ((k,v) -> m a) -> m [a]
-once xs = forM (Map.toList (Map.fromList xs))
+once xs action = fmap catMaybes $ flip evalStateT mempty $
+    forM xs $ \(k, v) -> do
+        s <- get
+        modify' (Set.insert k)
+        case Set.member k s of
```
Comment (paweljakubas) [inline lib/core/test/unit/Cardano/Wallet/DB/Properties.hs]: almost sure hlint will shout here and force you to use `if ... then ... else` :-)

File: lib/core/test/unit/Cardano/Wallet/DB/Arbitrary.hs (hunk 7 above)
Comment (KtorZ) [reply lib/core/test/unit/Cardano/Wallet/DB/Arbitrary.hs]: Indeed, it's the same constant as for the generated slot id (which is arbitrary). I'll use a named constant :+1:

File: lib/core/test/unit/Cardano/Wallet/DB/Arbitrary.hs (hunk 6 above)
Comment (KtorZ) [reply lib/core/test/unit/Cardano/Wallet/DB/Arbitrary.hs:228]: Indeed. It has proven to be sufficient long chain already to find quite a lot of bugs :sweat_smile:

File: lib/core/test/unit/Cardano/Wallet/DB/Properties.hs (hunk 9 above)
Comment (KtorZ) [reply lib/core/test/unit/Cardano/Wallet/DB/Properties.hs]: It's weird... I had quite a few problems with hlint today, seems like my editor doesn't run it anymore.. need to look into that.

**Comment by KtorZ** [comment]
@paweljakubas done :+1:

**Review by paweljakubas — APPROVED** [review]

**Comment by KtorZ** [comment]
bors retry
